contracts/chainlink/OracleHubChainLink.sol

```diff
@@ -0,0 +1,142 @@
+        setLinkToken(linkTokenAddress);
+        setOracle(oracleAddress);
+    }
+
+    /*
+    // EXTERNAL METHODS
+    */
+
+    /// @dev allows us to withdraw link as the owner of this contract
+    /// @param sendTo       address of where to send LINK
+    /// @param amount       amount of link (in baseUnits)
+    function withdrawLink(address sendTo, uint256 amount) external onlyOwner returns (bool) {
+        return link.transfer(sendTo, amount);
+    }
+
+    /// @notice called by a user requesting an on demand query when they believe that a contract is pushed into
+    /// an expired state.
+    /// @param marketContractAddress - address of the MarketContract
+    function requestOnDemandQuery(address marketContractAddress) external {
+        // TODO: enforce ETH payment to cover LINK, refund if settled?
+        ChainLinkQuery memory chainLinkQuery = contractAddressToChainLinkQuery[msg.sender];
+        bytes32 jobId;
+        createRunAndRequest(jobId, chainLinkQuery);
+        // if we refund for pushing to settlement we will need to record the caller of this function
+        // since it will be asynchronous with the callback
+    }
+
+    /// @dev allows for the owner to set a factory address that is then allowed the priviledge of creating
+    /// queries.
+    function setFactoryAddress(address contractFactoryAddress) external onlyOwner {
+        marketContractFactoryAddress = contractFactoryAddress;
+    }
+
+    /// @dev called by our factory to create the needed initial query for a new MarketContract
+    /// @param marketContractAddress - address of the newly created MarketContract
+    /// @param oracleQueryURL   URL of rest end point for data IE 'https://api.kraken.com/0/public/Ticker?pair=ETHUSD'
+    /// @param oracleQueryPath  path of data inside json object. IE 'result.XETHZUSD.c.0'
+    function requestQuery(
+        address marketContractAddress,
+        string oracleQueryURL,
+        string oracleQueryPath
+    ) external onlyFactory
+    {
+        ChainLinkQuery storage chainLinkQuery = contractAddressToChainLinkQuery[marketContractAddress];
+        chainLinkQuery.marketContractAddress = marketContractAddress;
+        chainLinkQuery.oracleQueryURL = oracleQueryURL;
+        chainLinkQuery.oracleQueryPath = oracleQueryPath;
+
+        // NOTE: currently this cannot be separated into its own function due to constraints around
+        // return types and the current version of solidity... is this true for internal / private fxs?
+        StringLib.slice memory pathSlice = oracleQueryPath.toSlice();
+        StringLib.slice memory delim = ".".toSlice();
+        chainLinkQuery.oracleQueryRunPath = new string[](pathSlice.count(delim) + 1);
+        for (uint i = 0; i < chainLinkQuery.oracleQueryRunPath.length; i++) {
+            chainLinkQuery.oracleQueryRunPath[i] = pathSlice.split(delim).toString();
+        }
+        bytes32 jobId; // TODO: how do we want to generate these?
```

Comment:
These will be given to you off-chain, and can change (I'll go into this detail below), so you may want to accept them as parameters when calling the method. This may get confusing, but I'll try to make sense: the actual JobID is a string that just so happens to be made up of hex characters, then when that's translated to a bytes32 value, it's a different hex value, from the conversion of the string to hex. For example, the JobID "ae18a03a2f5746ca967c403cf53e1318" would be "0x6165313861303361326635373436636139363763343033636635336531333138" as bytes32. For the most part, doing something like `bytes32("ae18a03a2f5746ca967c403cf53e1318")` should be fine, but I thought you should be aware. Remix doesn't like the original string input for bytes32 parameters, so you'd have to go to a site like [this](https://www.rapidtables.com/convert/number/ascii-hex-bin-dec-converter.html) to convert it first.

Now, a jobId can change, especially on a testnet, if we need to migrate the node to a new oracle contract address. This may happen from time-to-time, as we're still in development ourselves. On main net, this would change if you've started a new agreement with nodes to retrieve the data you want.

Reply:
@thodges-gh - thanks this is very helpful.  So would we have a single `jobId` for use with a segment of oracles that many contracts could be using?  

For rinkeby, when we deploy you will give us a jobId?

Reply:
@pelsasser Correct, you can share the jobId between contracts. On Rinkeby, I have the following JobID already set up for you (I think you used this one previously):

JobID: **af34cc57442542a89a00ceae096fe540**
Contains the sleep and multiplier adapters, will return uint256 type. If you don't supply an "until" key, the job will start immediately. Additional JobIDs are available [here](https://docs.chain.link/v1.0/docs/getting-started#section-job-specs) but these do not contain the sleep adapter.

If (and more likely _when_) that JobID changes, I'll let you know, and I'll try to give you a warning before-hand, just in case.